### PR TITLE
Optimize redundant BatchNorm Nodes in GraphOptimizer

### PR DESCRIPTION
**Pull request by ksaurabh-cadence**
Summary:

Documentation:
Optimize redundant BatchNorm Nodes in GraphOptimizer

[Optional Fixes #issue]

Test Plan:

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -2077,12 +2077,62 @@ bool normalizeWeights(Module *M, ConvolutionNode &CV,
...
+      }
+      float eps = BN->getEpsilon();
+
+      // Redundancy check based on strict equalities.
+      if (scale == 1.0f && bias == 0.0f && mean == 0.0f &&
+          (var + eps == 1.0f)) {
+        BN->getResult().replaceAllUsesOfWith(BN->getInput());
+        changed = true;
+        continue;
+      }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2123]: This should be covered by the relaxed redundancy check below, right? Do we need it?

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -2077,12 +2077,62 @@ bool normalizeWeights(Module *M, ConvolutionNode &CV,
...
   LOG_SCOPE(F->getLogContext(), getName());
   bool changed = false;
   auto &nodes = F->getNodes();
   auto *M = F->getParent();
 
+  // For each node:
+  for (auto &node : nodes) {
+    // Remove BN if mean,var,eps,scale,beta values make it redundant as per
+    // expression (X - mean) * (1.0 / sqrt(var + eps)) * scale + bias.
+    if (auto *BN = dyn_cast<BatchNormalizationNode>(&node)) {
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2100]: nit to reduce indentation
```suggestion
    auto *BN = dyn_cast<BatchNormalizationNode>(&node);
    if (!BN) {
      continue;
    }
```

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -2077,12 +2077,62 @@ bool normalizeWeights(Module *M, ConvolutionNode &CV,
...
+      // B = (bias - mean * (1.0 / sqrt(var + eps)) * scale)
+      if (bias == mean && (sqrt(var + eps) == scale)) {
+        BN->getResult().replaceAllUsesOfWith(BN->getInput());
+        changed = true;
+      }
+    }
+  } // For all nodes in the graph.
+
   // For each node:
   for (auto &node : nodes) {
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2137]: This loop is still going to process BNs that are dead but haven't been removed by DCE. Either we should merge the new optimization loop you added into this one, or else move this into a separate pass from the existing pass (would prefer not to call DCE in the middle of the pass explicitly).

**Review by jfix71 — COMMENTED** [review]
Cool! Added a few comments.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (ksaurabh-cadence) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2140]: Good point! Agree, will merge the loops.

**Comment by ksaurabh-cadence** [comment]
@jfix71 All comments addressed! Thanks for quick reviews, appreciate it!

**Comment by ksaurabh-cadence** [comment]
@jfix71 can we merge this?

**Review by jfix71 — APPROVED** [review]
🚢

**Pull request closed without merging**